uarray/_uarray_dispatch.cxx

```diff
@@ -917,6 +1185,99 @@ PyObject * Function::get_replacer(Function * self) {
+    0,                                 /* tp_getattr */
+    0,                                 /* tp_setattr */
+    0,                                 /* tp_reserved */
+    0,                                 /* tp_repr */
+    0,                                 /* tp_as_number */
+    0,                                 /* tp_as_sequence */
+    0,                                 /* tp_as_mapping */
+    0,                                 /* tp_hash  */
+    0,                                 /* tp_call */
+    0,                                 /* tp_str */
+    0,                                 /* tp_getattro */
+    0,                                 /* tp_setattro */
+    0,                                 /* tp_as_buffer */
+    Py_TPFLAGS_DEFAULT,                /* tp_flags */
+    0,                                 /* tp_doc */
+    0,                                 /* tp_traverse */
+    0,                                 /* tp_clear */
+    0,                                 /* tp_richcompare */
+    0,                                 /* tp_weaklistoffset */
+    0,                                 /* tp_iter */
+    0,                                 /* tp_iternext */
+    BackendState_Methods,              /* tp_methods */
+    0,                                 /* tp_members */
+    0,                                 /* tp_getset */
+    0,                                 /* tp_base */
+    0,                                 /* tp_dict */
+    0,                                 /* tp_descr_get */
+    0,                                 /* tp_descr_set */
+    0,                                 /* tp_dictoffset */
+    0,                                 /* tp_init */
+    0,                                 /* tp_alloc */
+    BackendState::new_,                /* tp_new */
+};
+
+PyObject * get_state(PyObject * /* self */, PyObject * /* args */) {
+  py_ref new_tuple = py_ref::steal(PyTuple_New(0));
+  if (!new_tuple)
+    return nullptr;
+
+  py_ref ref = py_ref::steal(PyObject_Call(
+      reinterpret_cast<PyObject *>(&BackendStateType), new_tuple.get(), NULL));
+  BackendState * output = reinterpret_cast<BackendState *>(ref.get());
+
+  output->locals = local_state;
+  output->use_thread_local_globals = use_thread_local_globals;
+
+  if (use_thread_local_globals)
+    output->globals = global_state;
+  else
+    output->globals = thread_global_state;
+
+  return ref.release();
+}
+
+PyObject * set_state(PyObject * /* self */, PyObject * args) {
+  PyObject * arg;
+  int reset_allowed = false;
+  if (!PyArg_ParseTuple(args, "O|p", &arg, &reset_allowed))
+    return nullptr;
+
+  if (!PyObject_IsInstance(
+          arg, reinterpret_cast<PyObject *>(&BackendStateType))) {
+    PyErr_SetString(
+        PyExc_TypeError, "state must be a uarray._BackendState object.");
+    return nullptr;
+  }
+
+  BackendState * state = reinterpret_cast<BackendState *>(arg);
+  local_state = state->locals;
+  global_state = state->globals;
```

Comment:
If I have multiple levels of `ua.set_state` then this is definitely wrong. You should be writing to the global state that is about to be activated
```cpp
// update use_thread_local_globals first, then

if (use_thread_local_globals) {
  thread_global_state = state->globals;
} else {
  global_state = state->globals;
}
```

Although, I would note that `global_state = state->globals` is still very dangerous in a multi-threaded environment. For the case of `with ua.set_state()` it's probably not what you want anyway. Just setting `use_thread_local_globals` to `false` should be enough to undo the actions from inside the `ua.set_state()` block.

Reply:
> Although, I would note that `global_state = state->globals` is still very dangerous in a multi-threaded environment.

We're not releasing the GIL anywhere in our code, so we should be fine.



> For the case of `with ua.set_state()` it's probably not what you want anyway. Just setting `use_thread_local_globals` to `false` should be enough to undo the actions from inside the `ua.set_state()` block.

This won't work for nested `ua.set_state`s.

Reply:
>We're not releasing the GIL anywhere in our code, so we should be fine.

The problem isn't necessarily in concurrent access to the variable but with changing the state in the middle of some function in another thread. Also note that while we don't release the GIL, we can still lose it when calling other python code. So, right in the middle of the dispatch process when we call `__ua_function__` for example.

>This won't work for nested ua.set_states.

Why wouldn't it? `ua.set_state` should only change the local state and the global state can't have been touched by the current thread.

Reply:
> Why wouldn't it? `ua.set_state` should only change the `thread_local` state and the shared global state can't have been touched by the current thread.

Because of nested fork-joins. It's not necessarily one level of hierarchy. In any case, I'm not writing to the global state anymore, as far as I can see.